sys/cheri/cheric.h

```diff
@@ -92,6 +91,27 @@
 	(cheri_getbase(c) + cheri_getlen(c));			\
 })
 
+#ifdef NOTYET
+#define	cheri_is_subset(x, y)	__builtin_cheri_subset_test(x, y)
+#else
+/*
+ * Soft implementation of cheri_subset_test(). To be replaced by LLVM intrinsic.
+ * Test whether a capability is a subset of another.
+ */
+#ifdef __cplusplus
+static inline bool
+#else
+static inline _Bool
+#endif
+cheri_is_subset(const void * __capability parent, const void * __capability ptr)
+{
+	return (cheri_gettag(parent) == cheri_gettag(ptr) &&
+	    cheri_getbase(ptr) >= cheri_getbase(parent) &&
+	    cheri_gettop(ptr) <= cheri_gettop(parent) &&
+	    (cheri_getperm(ptr) & cheri_getperm(parent)) == cheri_getperm(ptr));
+}
```

Comment:
Does this need to be polymorphic or are our only uses (kernel and mrs) passing `void * __capability`'s?

Reply:
Good point. The kernel always uses it with pointers but I'm not sure in general. I'll make it a macro to be safe.